flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DefaultJobManagerRunnerRegistryTest.java

```diff
@@ -157,25 +144,33 @@ void testFailingLocalCleanup() {
 
         assertThatFuture(
                         testInstance.localCleanupAsync(
-                                jobManagerRunner.getJobID(), Executors.directExecutor()))
+                                jobManagerRunner.getJobID(),
+                                Executors.directExecutor(),
+                                new TestComponentMainThreadExecutor(Thread.currentThread())))
                 .isCompletedExceptionally()
                 .eventuallyFailsWith(ExecutionException.class)
                 .extracting(FlinkAssertions::chainOfCauses, FlinkAssertions.STREAM_THROWABLE)
                 .hasExactlyElementsOfTypes(ExecutionException.class, expectedException.getClass())
                 .last()
                 .isEqualTo(expectedException);
-        assertThat(testInstance.isRegistered(jobManagerRunner.getJobID())).isFalse();
+        assertThat(testInstance.isRegistered(jobManagerRunner.getJobID()))
+                .as(
+                        "Since the cleanup failed, the JobManagerRunner is expected to not have been unregistered.")
+                .isTrue();
     }
 
     @Test
-    void testSuccessfulLocalCleanupAsync() {
```

Comment:
If we overload the method `registerTestingJobManagerRunner`:
```java
    private TestingJobManagerRunner registerTestingJobManagerRunner() {
        final TestingJobManagerRunner jobManagerRunner = TestingJobManagerRunner.newBuilder().build();
        registerTestingJobManagerRunner(jobManagerRunner);

        return jobManagerRunner;
    }

    private void registerTestingJobManagerRunner(
            TestingJobManagerRunner jobManagerRunner) {
        testInstance.register(jobManagerRunner);

        assertThat(testInstance.isRegistered(jobManagerRunner.getJobID())).isTrue();
        assertThat(jobManagerRunner.getTerminationFuture()).isNotDone();
    }
```
we can make this test even more reactive:

```java
    @RegisterExtension
    static final TestExecutorExtension<ScheduledExecutorService> EXECUTOR_EXTENSION =
            TestingUtils.defaultExecutorExtension();
    // [...]
    @Test
    void testSuccessfulLocalCleanupWaitsForDeregistration() {
        final ManuallyTriggeredComponentMainThreadExecutor mainThreadExecutor =
                new ManuallyTriggeredComponentMainThreadExecutor(Thread.currentThread());
        final TestingJobManagerRunner jobManagerRunner = registerTestingJobManagerRunner();
        final CompletableFuture<Void> cleanupResult =
                testInstance.localCleanupAsync(
                        jobManagerRunner.getJobID(),
                        Executors.directExecutor(),
                        mainThreadExecutor);

        jobManagerRunner.getTerminationFuture().complete(null);

        FlinkAssertions.assertThatFuture(cleanupResult)
                .as("Wait for the de-registration to complete")
                .isNotDone();

        mainThreadExecutor.triggerAll();

        FlinkAssertions.assertThatFuture(cleanupResult).eventuallySucceeds();
    }

    @Test
    void testSuccessfulLocalCleanupReliesOnJobManagerRunnerClose() {
        final TestingJobManagerRunner jobManagerRunner =
                TestingJobManagerRunner.newBuilder().setBlockingTermination(true).build();
        registerTestingJobManagerRunner(jobManagerRunner);
        final CompletableFuture<Void> cleanupResult =
                CompletableFuture.runAsync(
                        () ->
                                testInstance.localCleanupAsync(
                                        jobManagerRunner.getJobID(),
                                        Executors.directExecutor(),
                                        // execute de-registration right-away w/o any additional
                                        // blocking
                                        ComponentMainThreadExecutorServiceAdapter.forMainThread()),
                        // TestingJobManagerRunner doesn't close asynchronously, so we have to run
                        // the close call in a separate thread to unblock the test main thread
                        EXECUTOR_EXTENSION.getExecutor());

        FlinkAssertions.assertThatFuture(cleanupResult)
                .as("Wait for close call to complete")
                .isNotDone();

        jobManagerRunner.getTerminationFuture().complete(null);

        FlinkAssertions.assertThatFuture(cleanupResult).eventuallySucceeds();
    }
```

I didn't run the code. So it might need some adjustment.